File: pyshell/utils/execution/foreground.py
import os
import utils.global_var as gv
import termios

"""
Functions to manipulate the controlling terminal directly.
Always open /dev/tty to perform manipulation on the controlling terminal
"""

def set_foreground(pgid):
    """
    Use to be called by a the a shell to give
    the foreground to any pgid, including himself.
    """
    with open("/dev/tty") as control_term:
        os.tcsetpgrp(control_term.fileno(), pgid)

def restore_tcattr():
    """
    Restore the termios configuration, usually called
    when the shell retrieve foreground.
    """
    with open("/dev/tty") as control_term:
        termios.tcsetattr(control_term.fileno(), termios.TCSADRAIN, \
            gv.TCSETTINGS)

def get_tpgid():
    """
    Return the current foreground process group.
    """
    with open("/dev/tty") as control_term:
        return os.tcgetpgrp(control_term.fileno())

File: src/execution/foreground.c
#include <unistd.h>
#include <fcntl.h>
#include "tmpsh.h"

/*
** set_foreground:
**
** @pgid: pgid to push in foreground.
*/

void	set_foreground(pid_t pgid)
{
	int		term_fd;

	term_fd = open("/dev/tty", O_RDONLY);
	tcsetpgrp(term_fd, pgid);
	close(term_fd);
}

/*
** restore_tcattr:
**
** Reset the terminal control's config to the initial state
** when the shell was launched, using the global variable g_tcsettings.
*/

void	restore_tcattr(void)
{
	int		term_fd;

	term_fd = open("/dev/tty", O_RDONLY);
	tcsetattr(term_fd, TCSADRAIN, &g_tcsettings);
	close(term_fd);
}

/*
** get_tpgid:
**
** return : - pgid of the foreground process group.
*/

int		get_tpgid(void)
{
	int		term_fd;
	int		tpgid;

	term_fd = open("/dev/tty", O_RDONLY);
	tpgid = tcgetpgrp(term_fd);
	close(term_fd);
	return (tpgid);
}
